src/h3lib/lib/polygon.c

```diff
@@ -93,3 +95,172 @@ bool pointInsidePolygon(const GeoPolygon* geoPolygon, const BBox* bboxes,
+ * @param  polygons     Polygons to test
+ * @param  polygonCount Number of polygons in the test array
+ * @return              Number of polygons containing the loop
+ */
+static int countContainers(LinkedGeoLoop* loop, LinkedGeoPolygon** polygons,
+                           BBox** bboxes, int polygonCount) {
+    int containerCount = 0;
+    for (int i = 0; i < polygonCount; i++) {
+        if (pointInsideLinkedGeoLoop(polygons[i]->first, bboxes[i],
+                                     &loop->first->vertex)) {
+            containerCount++;
+        }
+    }
+    return containerCount;
+}
+
+/**
+ * Find the polygon to which a given hole should be allocated. Note that this
+ * function will return null if no parent is found.
+ * @param  loop         Inner loop describing a hole
+ * @param  polygon      Head of a linked list of polygons to check
+ * @param  polygonCount Number of polygons to check
+ * @return              Pointer to parent polygon, or null if not found
+ */
+static LinkedGeoPolygon* findPolygonForHole(LinkedGeoLoop* loop,
+                                            LinkedGeoPolygon* polygon,
+                                            BBox* bboxes, int polygonCount) {
+    // Early exit with no polygons
+    if (polygonCount == 0) {
+        return NULL;
+    }
+    // Initialize arrays for candidate loops and their bounding boxes
+    LinkedGeoPolygon** candidates =
+        calloc(polygonCount, sizeof(LinkedGeoPolygon*));
+    assert(candidates != NULL);
+    BBox** candidateBBoxes = calloc(polygonCount, sizeof(BBox*));
+    assert(candidateBBoxes != NULL);
+
+    // Find all polygons that contain the loop
+    int containerCount = 0;
+    int index = 0;
+    while (polygon) {
+        // We are guaranteed not to overlap, so just test the first point
+        if (pointInsideLinkedGeoLoop(polygon->first, &bboxes[index],
+                                     &loop->first->vertex)) {
+            candidates[containerCount] = polygon;
+            candidateBBoxes[containerCount] = &bboxes[index];
+            containerCount++;
+        }
+        polygon = polygon->next;
+        index++;
+    }
+
+    // Set the initial return value to the first candidate
+    LinkedGeoPolygon* parent = candidates[0];
+
+    // If we have multiple candidates, they must be nested inside each other.
+    // Find the innermost polygon by taking the candidate with the most
+    // containers in the candidate list.
+    if (containerCount > 1) {
+        int max = -1;
+        for (int i = 0; i < containerCount; i++) {
+            int count = countContainers(candidates[i]->first, candidates,
+                                        candidateBBoxes, containerCount);
+            if (count > max) {
+                parent = candidates[i];
+                max = count;
+            }
+        }
+    }
+
+    // Free allocated memory
+    free(candidates);
+    free(candidateBBoxes);
+
+    return parent;
+}
```

Comment:
This function could use a few helpers to clarify what is going on. I'm losing the flow of it due to the length.

Reply:
I could probably break out the second loop into a function. I'll take a look.

Reply:
Thanks.